AssertionError raised after #19248

<!--
If you're not sure whether what you're experiencing is a mypy bug, please see the "Question and Help" form instead.
Please consider:

- checking our common issues page: https://mypy.readthedocs.io/en/stable/common_issues.html
- searching our issue tracker: https://github.com/python/mypy/issues to see if it's already been reported
- asking on gitter chat: https://gitter.im/python/typing
-->

**Bug Report**

<!--
If you're reporting a problem with a specific library function, the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues

If the project you encountered the issue in is open source, please provide a link to the project.
-->

After upgrading to 1.16.1 we are seeing occasional `AssertionError`s being logged:

```console
$ mypy
mypy 1.16.1 (compiled: yes)
Exception ignored in PyObject_HasAttr(); consider using PyObject_HasAttrWithError(), PyObject_GetOptionalAttr() or PyObject_GetAttr():
Traceback (most recent call last):
  File "mypy/nodes.py", line 615, in setter
AssertionError: 
... (repeated many times)
```

It seems like the code introduced in https://github.com/python/mypy/pull/19248 to fix https://github.com/python/mypy/issues/19224 makes an assert which is not always true. It seems like the worry expressed in a [comment](https://github.com/python/mypy/pull/19248#pullrequestreview-2907639889) on that PR was justified ;).

**To Reproduce**

We haven't managed to find a reliable reproducer yet. Switching between branches in our repo and invoking `mypy`, sometimes triggers it, but sometimes doesn't.

**Expected Behavior**

<!--
How did you expect mypy to behave? It’s fine if you’re not sure your understanding is correct.
Write down what you thought would happen. If you expected no errors, delete this section.
-->

No errors.

**Actual Behavior**

<!-- What went wrong? Paste mypy's output. -->

Errors.

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used: 1.6.1
- Mypy command-line flags: `dmypy run -- --cache-dir .mypy_cache/not_recently_touched <explicitly specified list of .py files>`
- Mypy configuration options from `mypy.ini` (and other config files):
```toml
[tool.mypy]
plugins = [
  "pydantic.mypy",
  "pydantic.v1.mypy",
  "sqlalchemy.ext.mypy.plugin",
  "mypy.plugins.proper_plugin", 
  <a custom mypy plugin>
]
python_version = "3.13"
show_error_codes = true
strict = true

enable_error_code = [
  "ignore-without-code",
  "redundant-expr",
  "redundant-self",
  "truthy-bool",
  "truthy-iterable",
]
no_implicit_optional = true
warn_unreachable = true

[[tool.mypy.overrides]]
ignore_missing_imports = true
module = [
    "asyncpg.*",
]
```
- Python version used:

```console
$ python --version
Python 3.13.3
```

I suspect the `astmerge` in the `dmypy` server triggers this error, because `OverloadedFuncDef.setter` is defined as a `@property`.

From the error logs above we can see that the `AssertionError` is triggered during an `hasattr` call. Searching the codebase for an interesting `hasattr`, we find an interesting `hasattr` call in [`replace_object_state`](https://github.com/python/mypy/blob/master/mypy/util.py#L400-L403). This results in possibly calling `hasattr` on `OverloadedFuncDef.setter`, which evaluates the property unconditionally (since `hasattr` is implemented using `getattr` [[1]](https://docs.python.org/3/library/functions.html#hasattr)), triggering the `assert` on [`mypy/nodes.py:615`](https://github.com/python/mypy/blob/v1.16.1/mypy/nodes.py#L615).

Main uses of `replace_object_state` are in [`mypy/server/astmerge.py`](https://github.com/python/mypy/blob/v1.16.1/mypy/server/astmerge.py).

Assuming the above I tried to create a reproducer:

```console
$ mkdir /tmp/hello
$ cd /tmp/hello
$ uv init
$ uv add mypy==1.16.1
$ cat >test.py <<EOL
from typing import Literal, overload

@overload
def a(x: Literal[1]) -> int: ...

@overload
def a(x: Literal["a"]) -> str: ...

@overload
def a(x: Literal["b"]) -> str: ...

def a(x: Literal[1, "a", "b"]) -> int | str:
    return x
EOL
$ uv run dmypy run test.py
Daemon started
Success: no issues found in 1 source file
$ cat >test.py <<EOL
from typing import Literal, overload

@overload
def a(x: Literal[1]) -> int: ...

@overload
def a(x: Literal["a"]) -> str: ...

def a(x: Literal[1, "a"]) -> int | str:
    return x
EOL
$ uv run dmypy run test.py
Exception ignored in PyObject_HasAttr(); consider using PyObject_HasAttrWithError(), PyObject_GetOptionalAttr() or PyObject_GetAttr():
Traceback (most recent call last):
  File "mypy/nodes.py", line 615, in setter
AssertionError: 
Exception ignored in PyObject_HasAttr(); consider using PyObject_HasAttrWithError(), PyObject_GetOptionalAttr() or PyObject_GetAttr():
Traceback (most recent call last):
  File "mypy/nodes.py", line 615, in setter
AssertionError: 
Exception ignored in PyObject_HasAttr(); consider using PyObject_HasAttrWithError(), PyObject_GetOptionalAttr() or PyObject_GetAttr():
Traceback (most recent call last):
  File "mypy/nodes.py", line 615, in setter
AssertionError: 
Exception ignored in PyObject_HasAttr(); consider using PyObject_HasAttrWithError(), PyObject_GetOptionalAttr() or PyObject_GetAttr():
Traceback (most recent call last):
  File "mypy/nodes.py", line 615, in setter
AssertionError: 
Success: no issues found in 1 source file

$ ... cycling between the versions of `test.py` consistently reproduces the error.
```

We probably want `OverloadedFuncDef.setter` to not be a `@property`, s.t. it is not evaluated by `hasattr`? Or we should prevent it from being evaluated by accident (e.g. in `replace_object_state`).

Quite a tricky bug ;).

--- a/mypy/server/astmerge.py
+++ b/mypy/server/astmerge.py
@@ -351,6 +351,9 @@ class NodeReplaceVisitor(TraverserVisitor):
                 # in external types (e.g. named tuples), so we need to update it manually.
                 skip_slots = ("special_alias",)
                 replace_object_state(new.special_alias, node.special_alias)
+            if isinstance(node, OverloadedFuncDef) and isinstance(new, OverloadedFuncDef):
+                # Non-settable property that can raise AssertionError's during patching
+                skip_slots = ("setter",)
             replace_object_state(new, node, skip_slots=skip_slots)
             return cast(SN, new)
         return node

--- a/mypy/server/astmerge.py
+++ b/mypy/server/astmerge.py
@@ -345,16 +345,11 @@ class NodeReplaceVisitor(TraverserVisitor):
     def fixup(self, node: SN) -> SN:
         if node in self.replacements:
             new = self.replacements[node]
-            skip_slots: tuple[str, ...] = ()
             if isinstance(node, TypeInfo) and isinstance(new, TypeInfo):
                 # Special case: special_alias is not exposed in symbol tables, but may appear
                 # in external types (e.g. named tuples), so we need to update it manually.
-                skip_slots = ("special_alias",)
                 replace_object_state(new.special_alias, node.special_alias)
-            if isinstance(node, OverloadedFuncDef) and isinstance(new, OverloadedFuncDef):
-                # Non-settable property that can raise AssertionError's during patching
-                skip_slots = ("setter",)
-            replace_object_state(new, node, skip_slots=skip_slots)
+            replace_object_state(new, node, skip_slots=_get_ignored_slots(new))
             return cast(SN, new)
         return node
 
@@ -559,9 +554,16 @@ def replace_nodes_in_symbol_table(
             if node.node in replacements:
                 new = replacements[node.node]
                 old = node.node
-                # Needed for TypeInfo, see comment in fixup() above.
-                replace_object_state(new, old, skip_slots=("special_alias",))
+                replace_object_state(new, old, skip_slots=_get_ignored_slots(new))
                 node.node = new
             if isinstance(node.node, (Var, TypeAlias)):
                 # Handle them here just in case these aren't exposed through the AST.
                 node.node.accept(NodeReplaceVisitor(replacements))
+
+
+def _get_ignored_slots(node: SymbolNode) -> tuple[str, ...]:
+    if isinstance(node, OverloadedFuncDef):
+        return ("setter",)
+    if isinstance(node, TypeInfo):
+        return ("special_alias",)
+    return ()

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -222,6 +222,14 @@ addopts = "-nauto --strict-markers --strict-config"
 # treat xpasses as test failures so they get converted to regular tests as soon as possible
 xfail_strict = true
 
+# Force warnings as errors
+filterwarnings = [
+  "error",
+  # Some testcases may contain code that emits SyntaxWarnings, and they are not yet
+  # handled consistently in 3.14 (PEP 765)
+  "default::SyntaxWarning",
+]
+
 [tool.coverage.run]
 branch = true
 source = ["mypy"]

